Firmware/LowLevel/src/main.cpp

```diff
@@ -140,66 +156,59 @@ void setRaspiPower(bool power) {
 }
 
 void updateEmergency() {
-
     if (millis() - last_heartbeat_millis > HEARTBEAT_MILLIS) {
         emergency_latch = true;
         ROS_running = false;
     }
     uint8_t last_emergency = status_message.emergency_bitmask & LL_EMERGENCY_BIT_LATCH;
-
-    // Read & assign emergencies in the same manner as in ll_status.emergency_bitmask
-    uint8_t emergency_read = !gpio_get(PIN_EMERGENCY_3) << 1 | // Stop1
-                             !gpio_get(PIN_EMERGENCY_4) << 2 | // Stop2
-                             !gpio_get(PIN_EMERGENCY_1) << 3 | // Lift1
-                             !gpio_get(PIN_EMERGENCY_2) << 4 | // Lift2
-                             stock_ui_emergency_state; // OR with StockUI emergency
     uint8_t emergency_state = 0;
 
-    // Some mowers have the logic level inverted, fix this...
-    emergency_read = emergency_read ^ (LL_EMERGENCY_BIT_LIFT1 * LIFT1_IS_INVERTED);
-    emergency_read = emergency_read ^ (LL_EMERGENCY_BIT_LIFT2 * LIFT2_IS_INVERTED);
+    // Check all "stop" mode emergency inputs (halls)
+    bool stop_pressed = false;
+    size_t i;
+    for (i = 0; i < MAX_HALL_INPUTS; i++) {
+        if (stop_halls[i] == 0xff) break;                                                            // End marker reached
+        stop_pressed = halls[stop_halls[i]]() ^ llhl_config.hall_configs[stop_halls[i]].active_low;  // Get hall value and apply active_low level (invert)
+        if (stop_pressed) break;                                                                     // Break at first detected stop
+    }
 
     // Handle emergency "Stop" buttons
-    if (emergency_read & LL_EMERGENCY_BITS_STOP) {
-        // If we just pressed, store the timestamp
-        if (button_emergency_started == 0) {
-            button_emergency_started = millis();
-        }
+    if (stop_pressed) {
+        if (button_emergency_started == 0) button_emergency_started = millis();  // If we just pressed, store the timestamp
     } else {
-        // Not pressed, reset the time
-        button_emergency_started = 0;
+        button_emergency_started = 0;  // Not pressed, reset the time
```

Comment:
Well the comments are in since about July 2022 :innocent: 
Don't know if Clemens integrated them or someone else.
In any case, I mostly try to not remove any comments from someone else.
Instead of, if I've the impression that the code is pretty clear, I move them to the end of line. By this, the one who struggles with the code, can see a short hint at the end, but it doesn't waste vertical code reading.